nixos/node-red: add module

MatthewCroughan opened the pull request:
Adds a basic nixos module/service for node-red based on nodePackages.node-red

Thanks to @nixinator for helping out a lot with this.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Node-RED has a package with no module/service, this adds a basic service that I hope to expand upon the options of in the future.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented:
You forgot to add it to the file that collects all modules.

MatthewCroughan committed:
nixos/node-red: add module

Adds a basic nixos module/service for node-red based on nodePackages.node-red

Co-authored-by: Aaron Andersen <aaron@fosslib.net>
Co-authored-by: Sandro <sandro.jaeckel@gmail.com>
Co-authored by: Adrian Parvin Ouano <adrianparvino@gmail.com>
Co-authored-by: Norbert Melzer <nmelzer@nobbz.dev>

nixinator commented:
> You forgot to add it to the file that collects all modules.

I presume you mean module-list.nix

the Nixos manual doesn't mention https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix that it needs be updated , rather like top-level for packages. Maybe I/WE/SOMEONE should add it to the docs?

SuperSandro2000 commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+    };
+
+    port = mkOption {
+      type = types.int;
```
```suggestion
      type = types.port;
```

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+      description = "Node-RED package to use.";
+    };
+
+    configFile = mkOption {
```
I'm left wondering if we should rename `configFile` to `settings`, change the `type` to `(pkgs.formats.json {}).type`, and then generate a `settings.js` file based on that :thinking: This would eliminate the need for the extra `define` option and make the module more configurable ... and in a declarative way.

MatthewCroughan replied:
Maybe both? I am copying most of the layout of this module from the Cryptpad module, which itself copies from a lot of other Node services. 

Some people prefer to write their config files and specify a path to a config written in the DSL for their app, rather than writing their config entirely in Nix. What you are proposing with `settings` would make it so that it was only possible to configure the app via Nix, the config file could no longer be passed, correct?

aanderse replied:
Some module authors choose the include both a declarative method (`settings`), as well as an override (`configFile`). If that's something you're interested in then my comment can be interpreted as renaming the `define` option to `settings`.

Be aware that the downside of doing this is that many options stop having meaning and it can cause a confusing user experience which almost certainly requires a sysadmin to read the module source code to understand certain implications.

I think in this case `settings` might be sufficient because it maps easily to a nix attribute set. Very easy to convert config.

MatthewCroughan replied:
> Be aware that the downside of doing this is that many options stop having meaning
I've seen people say that too many options are a bad thing, yes. But it is fairly common to see a configFile option in Node modules, and other modules. This is the reason home-manager exists isn't it? This difference in opinion?

I would want both, to accommodate the people who don't want to deal with fully declarative configs and purity just yet. 

> rename configFile to settings, change the type to (pkgs.formats.json {}).type, and then generate a settings.js file based on that

Would I or would I not be able to simply pass a `path/to/config.json` if this were implemented? I want people to be able to just write and manage a config.json in their git repo, without having to interface 100% with Nix.

aanderse replied:
`builtins.fromJSON`?

I'm just offering suggestions on this, though. Please don't take my word as final or authoritative.

MatthewCroughan replied:
This is something that should be added by someone willing to submit he patch, since I don't feel strongly about how the config file is generated/parsed, the end result is more or less the same and it's a matter of improvement to be made to this basic module I'm providing. I think we should get it into Nix first so that it can be used, before worrying about how the configuration file is generated, although I think it's important that we provide these options sooner rather than later, I personally wouldn't have a need for it.

aanderse replied:
Sure :+1:

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+    userDir = mkOption {
+      type = types.path;
+      default = "/var/lib/node-red";
+      description = "the directory to store all user data, such as flow and credential files and all library data.";
```
```suggestion
      description = "The directory to store all user data, such as flow and credential files and all library data.";
```

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+      after = [ "networking.target" ];
+      serviceConfig = {
+        DynamicUser = true;
+        Environment = [
+          "HOME=${cfg.userDir}"
+        ];
```
Personally I prefer `systemd.services.node-red.environment.HOME = cfg.userDir;` here.

MatthewCroughan replied:
I prefer a list of values rather than an attrset of values since I think it's easier to read and reason about. What do you think?  It might just be personal preference at the end of the day. Are there any other reasons to use an attrset rather than readability?

aanderse replied:
A list cannot be overridden. If a sysadmin had good reason to they could write `systemd.services.node-red.environment.HOME = lib.mkForce "something-else";`.

MatthewCroughan replied:
Has the latest commit addressed this feedback? 669bcba

aanderse replied:
Love it :+1: Thanks.

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+        PrivateTmp = true;
+        Restart = "always";
+        StateDirectory = "node-red";
+        WorkingDirectory = "${cfg.userDir}";
```
```suggestion
        WorkingDirectory = cfg.userDir;
```

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+          "HOME=${cfg.userDir}"
+        ];
+        ExecStart = "${cfg.package}/bin/node-red --settings ${cfg.configFile} --port ${toString cfg.port} --userDir ${cfg.userDir} ${concatStringsSep " " (mapAttrsToList (name: value: "-D ${name}=${value}") cfg.define)}";
+        PrivateTmp = true;
```
This is already implied by `DynamicUser`.

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+      description = "List of settings.js overrides to pass via -D to Node-RED.";
+      example = literalExample ''
+        {
+          "logging.console.level=" = "trace";
```
```suggestion
          "logging.console.level" = "trace";
```

typo.

But see my above comment about possibly eliminating this option entirely.

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+
+    safe = mkOption {
+      type = types.bool;
+      default = "false";
```
```suggestion
      default = false;
```

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+      description = "Listening port.";
+    };
+
+    userDir = mkOption {
```
It doesn't seem like this is a valuable option. It seems like it will cause more problem than anything. What is going to happen if the sysadmin changes this value? In almost every scenario the application will likely fail to start, or start in a degraded state. I would suggest removing this.

MatthewCroughan replied:
> What is going to happen if the sysadmin changes this value?

What do you mean by this? Can you give me an example of sysadmins changing this value?
 
Most services in NixOS have the option to declare where the state is stored, why should this be any different?

MatthewCroughan replied:
Have I corrected this with this commit @aanderse?

cd7284d

aanderse replied:
`userDir` still isn't created if the user specifies anything other than the default, and `/var/lib/node-red` is created even if the user specifies something else. How about [this](https://github.com/NixOS/nixpkgs/commit/42219a0df086a580d96e54dbd65d6f081c6dc20a)?

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,79 @@
+      description = "the directory to store all user data, such as flow and credential files and all library data.";
+    };
+
+    safe = mkOption {
```
This option isn't used so you should remove it.

MatthewCroughan replied:
1c57642c6e66227cbb20ecc6e6a8f1180765ada0

The intent is to allow people to launch node-red in safe mode, if needed. This should solve it.

aanderse reviewed (changes requested):
Thanks for your contribution to nixpkgs! I have left a few comments I hope you find helpful.

I'm not familiar with `node-red` so personally I would feel much better with a NixOS test which demonstrates usage of this service and proves your module works on a going forward basis. Future PRs against this module, or against the `node-red` package could then be validated against your NixOS test.

MatthewCroughan commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,112 @@
+      default = defaultUser;
+      description = ''
+        Node-RED will be run under this user (user will be created if it doesn't exist.
+        This can be your user name).
```
Should we make it such that the user *is* created for you? Or is it correct to leave this to the system administrator?

aanderse replied:
Typical NixOS options would say something like `User under which Node-RED runs.` and you should probably follow that convention. I'll additionally note that I do not like the "This can be your user name" comment as this leads to the bad practice of users running system services as non system users.

I think all the NixOS `user` and `group` options should be updated to say something like:
```
User under which ____ runs. If left as the default value this user will automatically be created
on system activation, otherwise the sysadmin is responsible for ensuring the user exists.
```
... but that is probably best left for a separate PR :smile:

MatthewCroughan replied:
Has the latest commit addressed your feedback? 7801b79

aanderse replied:
LGTM :+1:

MatthewCroughan commented:
I think all that remains is for me to make a NixOS test in order to demonstrate usage of this module and also prove that it works. At which point we can merge, unless there are any more comments :smile:

SuperSandro2000 commented on 42219a0df086a580d96e54dbd65d6f081c6dc20a.patch:
```diff
@@ -0,0 +1,59 @@
+From 42219a0df086a580d96e54dbd65d6f081c6dc20a Mon Sep 17 00:00:00 2001
```
What is this file?

MatthewCroughan replied:
An accident :smile:

MatthewCroughan replied:
Fixed.

MatthewCroughan commented:
I have just squashed all of the commits to this point into a single commit which is coauthored by Sandro and Aaron.

MatthewCroughan committed:
nixos/node-red: add test

MatthewCroughan commented:
I've added a new option `withNpmAndGcc` which might be controversial. Essentially, node-red needs access to NPM and GCC at runtime in order to do its own package management, there is no real solution to this problem in Nix at this time, so I thought I'd add this option in order to make the program work as it's designed to be used for the time being. Would appreciate any comments. It might for example be more sensible to have a `node-red_withNpmAndGcc` rather than to have it be an option in the service, for example.

In addition, I've also added a NixOS Test which can be build like `nix build -f nixos/tests/node-red.nix -L` if you were in the root of the Nixpkgs git repo.

aanderse commented on nixos/modules/services/web-apps/node-red.nix:
```diff
@@ -0,0 +1,151 @@
+      '';
+    };
+
+    withNpmAndGcc = mkOption {
```
*Maybe* we replace this option with an `example` (and `description`) in the `package` option?
```
options.services.node-red.package = mkOption {
  default = pkgs.nodePackages.node-red;
  defaultText = "pkgs.nodePackages.node-red";
  type = types.package;
  description = ''
    Node-RED package to use.

    If you wish to give Node-RED access to NPM and GCC at runtime, so Nodes can be downloaded and managed
    via the 'Pallet Manager' you can use the example provided. Note that this will blah blah blah imperativeagainst
    blah blah blah work against the "nix way", etc...
  '';
  example = literalExample ''
    pkgs.nodePackages.node-red.override (oldArgs: {
      nativeBuildInputs = [ pkgs.makeWrapper];
      buildInputs = oldArgs.buildInputs ++ [ pkgs.nodePackages.npm pkgs.gcc ];
      postInstall = ''
        wrapProgram "$out/bin/node-red" --prefix PATH : ${
          pkgs.lib.makeBinPath [ pkgs.nodejs pkgs.gcc ]
        }
      '';
    })
  '';
};
```

Having this as an actual option seems like NixOS is promoting something that NixOS doesn't want to promote, so including a simple example seems more fitting in my humble opinion.

MatthewCroughan replied:
I disagree. This service is only useful with access to the library of 3000+ nodes available from https://flows.nodered.org/search?type=node.I think it's very bad to provide services that are not functional or practical at all. I would only support the addition of Node-RED as a service to nixpkgs with this option remaining present.

e5bc2a3 makes the `withNpmAndGcc` option wrap any `cfg.package` 
4ab3d42 updates the description in line with what you're suggesting, but only as much as necessary.

MatthewCroughan replied:
Unless you plan to add all 3000 nodes yourself to Nixpkgs and keep them up to date, I don't really see an alternative to allowing this option to exist. Otherwise, the service should not be in Nix at all, and I'll just provide it as a flake, if Nixpkgs is really this hostile to functionality.

MatthewCroughan replied:
If I'm honest, this option should be set to true, since you won't really be able to do that much with Node-RED without access to the palette library.

aanderse replied:
> This service is only useful with access to the library of 3000+ nodes available from https://flows.nodered.org/search?type=node.

Thanks for the context. It sounds like the option doesn't make sense then... because it should always be true?

MatthewCroughan replied:
> Thanks for the context. It sounds like the option doesn't make sense then... because it should always be true?

No. Because someone could write an expression to include the nixpkgs using node2nix, or any number of methods of declaring the packages to be 100% declared.

In such a case, it would be very relevant to remove NPM and GCC, because if Node-RED detects that NPM/GCC are not present at runtime, it disables the internal palette manager.

This option exists for the majority of people, not the few that will express every out-of-nixpkgs npm dependency they want.

nixinator replied:
yep, until we can get ecosystem like node red to become nix way friendly, then stuff like can't be run natively on nix. A chicken and egg problem. Maybe we should lay and egg, make chicken, then ask the chicken to lay egg's differently.

MatthewCroughan replied:
> yep, until we can get ecosystem like node red to become nix way friendly, then stuff like can't be run natively on nix. A chicken and egg problem. Maybe we should lay the egg, make chicken, then ask the chicken to lay egg's differently.

It runs fine on Nix. What does native mean? There is no way we're ever going to incorporate other package ecosystems 100% into expressions, the problem is the underlying software. We cannot support something that wants to use a different package ecosystem, using the Nix language. All we can do is give those pieces of software access to whatever pieces of software they want at runtime, so they can do their thing.

MatthewCroughan replied:
We can automate some of that work. But that's a long way away, and I'd rather get software into nixpkgs instead of eternal bikeshedding about how declarative something is, as if it's the job of the software to change right now, just because we want to package it.

aanderse reviewed (commented):
Thanks for adding a test!

MatthewCroughan commented:
I have once again squashed the history and have added two new co-authors (Since I got quite a bit of help from various channels with this module).

In addition to the squash, I've added the test to `nixos/tests/all-tests.nix` which I had forgotten to do previously.

Now, anyone can reproduce the test like so:
`nix build github:matthewcroughan/nixpkgs/node-red-service#nixosTests.node-red` will get you building the test. Works great.

nixinator commented:
your flake test runs and completes.

aanderse reviewed (approved):
LGTM :+1:

Should I be clicking the merge button now? 😄

MatthewCroughan commented:
@aanderse Yeah, I think this is ready now.

aanderse commented:
Thanks @MatthewCroughan!
